### PR TITLE
Add API calls for getting all possible fees (resolves #852)

### LINKED ISSUE #852: Add API calls to get percent of the fees
Talking about 2 API calls, one for withdraw fee, another for exchange fee on each currency

### PR DESCRIPTION
**Pull request by shal**

### CODE REVIEW

**Comment by yivo** [comment]
@ashanaakh Don't forget about docs and specs.

File: app/api/api_v2/fees.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+module APIv2
+  class Fees < Grape::API
+    Fee         = Struct.new :type, :value
+    WithdrawFee = Struct.new :currency, :type, :fee
+    TradingFee  = Struct.new :market, :ask_fee, :bid_fee
+    
+    desc 'Returns currencies and their withdraw fees.'
+    get '/fees/withdraw' do
+      withdraw_fees = Currency.visible.map do |c|
+        fee = Fee.new 'fixed', c.withdraw_fee
```
Comment (yivo) [inline app/api/api_v2/fees.rb]: @ashanaakh Please wrap in `(` & `)`. In Ruby it is better to use symbols for `:fixed`.

File: app/api/api_v2/fees.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+module APIv2
+  class Fees < Grape::API
+    Fee         = Struct.new :type, :value
```
Comment (yivo) [inline app/api/api_v2/fees.rb]: @ashanaakh Please use `(` `)` more often because it adds readability.

File: app/api/api_v2/fees.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    desc 'Returns currencies and their withdraw fees.'
+    get '/fees/withdraw' do
+      withdraw_fees = Currency.visible.map do |c|
+        fee = Fee.new 'fixed', c.withdraw_fee
+        WithdrawFee.new c.code, c.type, fee
+      end
+      present withdraw_fees
+    end
+
+    desc 'Returns markets and their fees.'
```
Comment (yivo) [inline app/api/api_v2/fees.rb]: @ashanaakh Better to write `Returns trading fees for markets.`.

File: app/api/api_v2/fees.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+module APIv2
+  class Fees < Grape::API
+    Fee         = Struct.new :type, :value
+    WithdrawFee = Struct.new :currency, :type, :fee
+    TradingFee  = Struct.new :market, :ask_fee, :bid_fee
+    
+    desc 'Returns currencies and their withdraw fees.'
```
Comment (yivo) [inline app/api/api_v2/fees.rb]: @ashanaakh Better to write `Returns withdraw fees for currencies`.

File: spec/api/api_v2/fees_spec.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
+describe APIv2::Fees, type: :request do
+  describe 'GET /api/v2/fees/withdraw' do
+    it 'returns currencies withdraw fees' do
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh better to write `it returns withdraw fees for every visible currency`.

File: spec/api/api_v2/fees_spec.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      Currency.find_by_code(:usd).update(withdraw_fee: 0.99)
+
+      get '/api/v2/fees/withdraw'
+
+      result = JSON.parse(response.body)
+      currency = result.find do |c|
+        c['currency'] == 'usd'
+      end
+
+      expect(response).to be_success
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh This must be before reading response body.

File: spec/api/api_v2/fees_spec.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      expect(result.size).to eq 5
+    end
+
+    it 'returns correct currency withdraw fee' do
+      Currency.find_by_code(:usd).update(withdraw_fee: 0.99)
+
+      get '/api/v2/fees/withdraw'
+
+      result = JSON.parse(response.body)
+      currency = result.find do |c|
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh Write this in single line.

File: app/api/api_v2/fees.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        fee = Fee.new 'fixed', c.withdraw_fee
+        WithdrawFee.new c.code, c.type, fee
+      end
+      present withdraw_fees
+    end
+
+    desc 'Returns markets and their fees.'
+    get '/fees/trading' do
+      trading_fees = Market.visible.map do |m|
+        ask_fee = Fee.new 'relative', m.ask_fee
```
Comment (yivo) [inline app/api/api_v2/fees.rb]: @ashanaakh Send fees as strings: `m.ask_fee.to_s('F')`.

File: app/api/api_v2/fees.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        bid_fee = Fee.new 'relative', m.bid_fee
```
Comment (yivo) [inline app/api/api_v2/fees.rb]: @ashanaakh Send fees as strings: `m.ask_fee.to_s('F')`.

File: app/api/api_v2/fees.rb (hunk 1 above)
Comment (yivo) [inline app/api/api_v2/fees.rb]: @ashanaakh Send fees as strings: `m.ask_fee.to_s('F')`.

File: spec/api/api_v2/fees_spec.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    it 'returns currencies withdraw fees' do
+      get '/api/v2/fees/withdraw'
+
+      expect(response).to be_success
+      result = JSON.parse(response.body)
+      expect(result.size).to eq 5
+    end
+
+    it 'returns correct currency withdraw fee' do
+      Currency.find_by_code(:usd).update(withdraw_fee: 0.99)
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh You can't just make any uncontrolled changes to this in specs or they will randomly fail. Just drop this line and compare directly: `Currency.find_by_code!(:usd).withdraw_fee`.

File: spec/api/api_v2/fees_spec.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+      expect(currency.dig('currency')).to eq 'usd'
+      expect(currency.dig('type')).to eq 'fiat'
+      expect(currency.dig('fee', 'value')).to eq '0.99'
+      expect(currency.dig('fee', 'type')).to eq 'fixed'
+    end
+  end
+
+  describe 'GET /api/v2/fees/trading' do
+    it 'returns markets fees' do
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh Just write `it returns trading fees`.

File: spec/api/api_v2/fees_spec.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+      get '/api/v2/fees/trading'
+      expect(response).to be_success
+
+      result = JSON.parse(response.body)
+
+      expect(result.size).to eq 1
+    end
+
+    it 'returns correct currency withdraw fee' do
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh Invalid message. Write `it returns correct trading fees`.

File: spec/api/api_v2/fees_spec.rb [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+      get '/api/v2/fees/trading'
+
+      result = JSON.parse(response.body)
+
+      market = result.find do |c|
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh Use single line for this.

File: spec/api/api_v2/fees_spec.rb [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        c['market'] == 'btcusd'
+      end
+
+      expect(response).to be_success
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh The same problem: don't parse result before checking if request is OK.

**Review by yivo — CHANGES_REQUESTED** [review]

**Comment by shal** [comment]
@yivo Thanks for review. I will fix it shortly

File: spec/api/api_v2/fees_spec.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    end
+  end
+
+  describe 'GET /api/v2/fees/trading' do
+    it 'returns trading fees' do
+      get '/api/v2/fees/trading'
+      expect(response).to be_success
+
+      result = JSON.parse(response.body)
+
```
Comment (yivo) [inline spec/api/api_v2/fees_spec.rb]: @ashanaakh It is better to not put an extra line here. Better to group linked operations: parsing result and them using it.

File: app/api/api_v2/fees.rb [hunk 16]
```diff
@@ -0,0 +1,27 @@
+module APIv2
+  class Fees < Grape::API
+    Fee         = Struct.new(:type, :value)
+    WithdrawFee = Struct.new(:currency, :type, :fee)
+    TradingFee  = Struct.new(:market, :ask_fee, :bid_fee)
+
+
```
Comment (yivo) [inline app/api/api_v2/fees.rb]: @ashanaakh Extra line detected.

**Review by yivo — CHANGES_REQUESTED** [review]
@ashanaakh Please, add one more endpoint: `/api/v2/fees/deposit`. Use `Currency#deposit_fee` for getting the value. Please add specs. It will fail until we merge #915 – it is OK.

**Review by yivo — APPROVED** [review]

**Pull request merged**